File: sdk/python/hitechcloud/resources/ceph_s3.py
"""Ceph S3 resource (11 endpoints)"""

from .base import BaseResource

class CephS3Resource(BaseResource):
    """Ceph S3: S3 connection, credentials, usage, buckets, sub-users, key rotation"""

    def get_connection_info(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3 - Get S3 connection info"""
        return self._http.get(f"/api/service/{service_id}/s3")

    def get_credentials(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3/credentials - Get S3 credentials with secret key"""
        return self._http.get(f"/api/service/{service_id}/s3/credentials")

    def get_usage(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3/usage - Get S3 usage"""
        return self._http.get(f"/api/service/{service_id}/s3/usage")

    def get_metrics(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3/metrics - Get S3 metrics"""
        return self._http.get(f"/api/service/{service_id}/s3/metrics")

    def list_buckets(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3/buckets - List buckets"""
        return self._http.get(f"/api/service/{service_id}/s3/buckets")

    def create_bucket(self, service_id: int, bucket: str) -> dict:
        """POST /api/service/{id}/s3/buckets - Create bucket"""
        return self._http.post(f"/api/service/{service_id}/s3/buckets", {"bucket": bucket})

    def delete_bucket(self, service_id: int, bucket: str) -> dict:
        """DELETE /api/service/{id}/s3/buckets/{bucket} - Delete bucket"""
        return self._http.delete(f"/api/service/{service_id}/s3/buckets/{bucket}")

    def list_subusers(self, service_id: int) -> dict:
        """GET /api/service/{id}/s3/subusers - List sub-users"""
        return self._http.get(f"/api/service/{service_id}/s3/subusers")

    def create_subuser(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/s3/subusers - Create sub-user"""
        return self._http.post(f"/api/service/{service_id}/s3/subusers", kwargs)

    def delete_subuser(self, service_id: int, subuser: str) -> dict:
        """DELETE /api/service/{id}/s3/subusers/{subuser} - Delete sub-user"""
        return self._http.delete(f"/api/service/{service_id}/s3/subusers/{subuser}")

    def rotate_secret_key(self, service_id: int) -> dict:
        """POST /api/service/{id}/s3/key - Rotate secret key"""
        return self._http.post(f"/api/service/{service_id}/s3/key")

File: sdk/ruby/lib/hitechcloud/resources/ceph_s3.rb
module HiTechCloud
  module Resources
    class CephS3 < Base

        # Get S3 info
        def s3(id)
          @http.get("/api/service/%{id}/s3" % { id: id })
        end

        # Get credentials
        def credentials(id)
          @http.get("/api/service/%{id}/s3/credentials" % { id: id })
        end

        # Get usage
        def usage(id)
          @http.get("/api/service/%{id}/s3/usage" % { id: id })
        end

        # Get metrics
        def metrics(id)
          @http.get("/api/service/%{id}/s3/metrics" % { id: id })
        end

        # List buckets
        def buckets(id)
          @http.get("/api/service/%{id}/s3/buckets" % { id: id })
        end

        # Create bucket
        def create_bucket(id, data)
          @http.post("/api/service/%{id}/s3/buckets" % { id: id }, data)
        end

        # Delete bucket
        def delete_bucket(id, bucket)
          @http.delete("/api/service/%{id}/s3/buckets/%{bucket}" % { id: id, bucket: bucket })
        end

        # List subusers
        def subusers(id)
          @http.get("/api/service/%{id}/s3/subusers" % { id: id })
        end

        # Create subuser
        def create_subuser(id, data)
          @http.post("/api/service/%{id}/s3/subusers" % { id: id }, data)
        end

        # Delete subuser
        def delete_subuser(id, subuser)
          @http.delete("/api/service/%{id}/s3/subusers/%{subuser}" % { id: id, subuser: subuser })
        end

        # Rotate key
        def rotate_key(id, data)
          @http.post("/api/service/%{id}/s3/key" % { id: id }, data)
      end
    end
  end
end
